invokeai/frontend/web/src/features/nodes/util/graphBuilders/addHrfToGraph.ts

```diff
@@ -0,0 +1,210 @@
+  const destinationFields = [
+    'vae',
+    'control',
+    'ip_adapter',
+    'metadata',
+    'unet',
+    'positive_conditioning',
+    'negative_conditioning',
+  ];
+  const newEdges: Edge[] = [];
+
+  // Loop through the existing edges connected to DENOISE_LATENTS
+  graph.edges.forEach((edge: Edge) => {
+    if (
+      edge.destination.node_id === DENOISE_LATENTS &&
+      destinationFields.includes(edge.destination.field)
+    ) {
+      // Add a similar connection to DENOISE_LATENTS_HRF
+      newEdges.push({
+        source: {
+          node_id: edge.source.node_id,
+          field: edge.source.field,
+        },
+        destination: {
+          node_id: DENOISE_LATENTS_HRF,
+          field: edge.destination.field,
+        },
+      });
+    }
+  });
+  graph.edges = graph.edges.concat(newEdges);
+}
+
+// Adds the high-res fix feature to the given graph.
+export const addHrfToGraph = (
+  state: RootState,
+  graph: NonNullableGraph
+): void => {
+  const { vae } = state.generation;
+  const isAutoVae = !vae;
+
+  // Pre-existing (original) graph nodes.
+  const originalDenoiseLatentsNode = graph.nodes[
+    DENOISE_LATENTS
+  ] as DenoiseLatentsInvocation;
+  const originalNoiseNode = graph.nodes[NOISE] as NoiseInvocation;
+  const originalLatentsToImageNode = graph.nodes[
+    LATENTS_TO_IMAGE
+  ] as LatentsToImageInvocation;
+
+  // Scale height and width by hrfScale.
+  const hrfScale = state.generation.hrfScale;
+  const scaledHeight = originalNoiseNode?.height
+    ? originalNoiseNode.height * hrfScale
+    : undefined;
+  const scaledWidth = originalNoiseNode?.width
+    ? originalNoiseNode.width * hrfScale
+    : undefined;
+
+  // Add hrf information to the metadata accumulator.
+  const metadataAccumulator = graph.nodes[METADATA_ACCUMULATOR] as
+    | MetadataAccumulatorInvocation
+    | undefined;
+  if (metadataAccumulator) {
+    metadataAccumulator.height = scaledHeight;
+    metadataAccumulator.width = scaledWidth;
+  }
+
+  // Define new nodes.
+  // Denoise latents node to be run on upscaled latents.
+  const denoiseLatentsHrfNode: DenoiseLatentsInvocation = {
+    type: 'denoise_latents',
+    id: DENOISE_LATENTS_HRF,
+    is_intermediate: originalDenoiseLatentsNode?.is_intermediate,
+    cfg_scale: originalDenoiseLatentsNode?.cfg_scale,
+    scheduler: originalDenoiseLatentsNode?.scheduler,
+    steps: originalDenoiseLatentsNode?.steps,
+    denoising_start: 1 - state.generation.hrfStrength,
+    denoising_end: 1,
+  };
+
+  const rescaleLatentsNode: RescaleLatentsInvocation = {
+    id: RESCALE_LATENTS,
+    type: 'lresize',
+    width: scaledWidth,
+    height: scaledHeight,
+  };
+
+  // New higher resolution noise node.
+  const hrfNoiseNode: NoiseInvocation = {
+    type: 'noise',
+    id: NOISE_HRF,
+    seed: originalNoiseNode.seed,
+    width: scaledWidth,
+    height: scaledHeight,
+    use_cpu: originalNoiseNode.use_cpu,
+    is_intermediate: originalNoiseNode.is_intermediate,
+  };
```

Comment:
Consider edges from the `rescaleLatentsNode`'s width and height output, to the `hrfNoiseNode`'s width and height input

Reply:
Why should I put graph edges in when I could initialize them this way in the code? Is that just more robust to future changes? 

Part of why I implemented this in the linear flow was I could use python for some of the more basic arguments and arithmatic, ex. I'm not using arithmetic nodes for calculating the scaled height + width. Should I use a node for as many operations as possible?

Reply:
Yeah it's more resilient to typos, and prevents the noise tensor from ever being mismatched to the image latents tensor, which will cause the generation to error. That said, I think this graph logic is safe for now.

Reply:
makes sense